base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,354 @@
+}
+
+// Print a string, aligned to right.
+// MaxWidth is the width for printing.
+// return FALSE if number length is longer than MaxWidth
+BOOL PrintNum(LONGLONG Number, INT MaxWidth)
+{
+    // calculate the length needed to print.
+    INT PrintLength = 0;
+    LONGLONG Tmp = Number;
+
+    if (Tmp < 0)
+    {
+        Tmp = -Tmp;
+        PrintLength++;
+    }
+
+    do
+    {
+        Tmp /= 10;
+        PrintLength += 1;
+    } while (Tmp);
+
+    if (PrintLength > MaxWidth)
+    {
+        return FALSE;
+    }
+    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
+
+    return TRUE;
+}
+
+// Print memory size using appropriate unit, with comma-separated number, aligned to right.
+// MaxWidth is the width for printing.
+// StartingUnit is the minimum memory unit used, 0 for Byte, 1 for KB, 2 for MB ...
+// return FALSE when failed to format.
+BOOL PrintMemory(SIZE_T MemorySizeByte, INT MaxWidth, INT StartingUnit)
+{
+    WCHAR *MemoryUnit[] = {L"B", L"K", L"M", L"G", L"T", L"P"};
+
+    MaxWidth -= 2; // a space and a unit will occupy 2 length.
+    if (MaxWidth <= 0)
+    {
+        return FALSE;
+    }
+
+    SIZE_T MemorySize = MemorySizeByte;
+
+    for (INT i = 0; i < _countof(MemoryUnit); i++)
+    {
+        if (i >= StartingUnit)
+        {
+            // calculate the length needed to print.
+            INT PrintLength = 0;
+            SIZE_T Tmp = MemorySize;
+            do
+            {
+                Tmp /= 10;
+                PrintLength += 1;
+            } while (Tmp);
+
+            if (PrintLength + (PrintLength - 1) / 3 <= MaxWidth) // (PrintLength - 1) / 3 is the length comma will take.
+            {
+                // enough to hold
+
+                // print padding space
+                PrintSpace(MaxWidth - (PrintLength + (PrintLength - 1) / 3));
+
+                INT Mod = 1;
+                for (INT i = 0; i < PrintLength - 1; i++)
+                {
+                    Mod *= 10;
+                }
+
+                for (INT i = PrintLength - 1; i >= 0; i--)
+                {
+                    ConPrintf(StdOut, L"%d", MemorySize / Mod);
+                    MemorySize %= Mod;
+                    if (i && i % 3 == 0)
+                    {
+                        ConPrintf(StdOut, L",");
+                    }
+                    Mod /= 10;
+                }
+
+                // print the unit.
+                ConPrintf(StdOut, L" %ls", MemoryUnit[i]);
+
+                return TRUE;
+            }
+        }
+        
+        MemorySize >>= 10; // MemorySize /= 1024;
+    }
+
+    // out of MemoryUnit range.
+    return FALSE;
+}
+
+VOID PrintHeader(VOID)
+{
+    ConPrintf(StdOut, L"\n");
+
+    WCHAR lpstrImageName[HEADER_STR_MAXLEN];
+    WCHAR lpstrPID[HEADER_STR_MAXLEN];
+    WCHAR lpstrSession[HEADER_STR_MAXLEN];
+    WCHAR lpstrMemUsage[HEADER_STR_MAXLEN];
+
+    LoadStringW(GetModuleHandle(NULL), IDS_HEADER_IMAGENAME, lpstrImageName, HEADER_STR_MAXLEN);
```

Comment:
Use `GetModuleHandleW` and call it only once.

Reply:
That makes sense. already adopted, thanks! 🤣